java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,660 @@
+      return type.is("java.lang.String");
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol) {
+      return taintSourceFor(symbol, Collections.emptyList());
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol, List<TaintSource> arguments) {
+      if (symbol.isMethodSymbol()) {
+        MethodSymbol method = (MethodSymbol)symbol;
+        TaintSource methodTaintSource = tsf.taintedIif(fullyQualify(symbol), arguments);
+
+        if (arguments.stream().anyMatch(ts -> ts.canBeTainted())) {
+          callees.add(methodTaintSource);
+        }
+
+        if (isString(method.returnType().type())) {
+          return methodTaintSource;
+        } else {
+          return tsf.taintFree();
+        }
+      } else if (symbol.isVariableSymbol() && symbol.owner().isTypeSymbol()) {
+        // Fields are not supported
+        return tsf.taintFree();
+      } else {
+        // Local variables, re-assigned parameters, etc.
+        return taintSources.computeIfAbsent(
+          symbol,
+          s -> tsf.taintedIif(fullyQualify(s), arguments));
+      }
+    }
+
+    public void put(Symbol symbol, TaintSource ts) {
+      taintSources.put(symbol, ts);
+    }
+
+    @Override
+    public int hashCode() {
+      return src.hashCode() +
+        13 * tsf.hashCode() +
+        17 * taintSources.hashCode() +
+        19 * callees.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
+        return false;
+      }
+
+      TaintProgramState other = (TaintProgramState) obj;
+
+      return src.equals(other.src) &&
+        tsf.equals(other.tsf) &&
+        taintSources.equals(other.taintSources) &&
+        callees.equals(other.callees);
+    }
+
+  }
+
+  public static TaintSummary computeTaintConditions(ScannedFile src, CFG cfg) {
+    TaintSourceFactory tsf = new TaintSourceFactory();
+
+    TaintSource result = tsf.taintFree();
+    Set<TaintSource> callees = new HashSet<>();
+
+    boolean returnsVoid = cfg.methodSymbol().returnType().type().isVoid();
+
+    Multimap<Block, TaintProgramState> visited = HashMultimap.create();
+    Multimap<Block, TaintProgramState> workList = HashMultimap.create();
+
+    Block entry = cfg.entry();
+    TaintProgramState entryState = new TaintProgramState(src, tsf);
+    workList.put(entry, entryState);
+
+    while (!workList.isEmpty()) {
+      Block block = workList.keys().iterator().next();
+      TaintProgramState state = workList.get(block).iterator().next();
+      visited.put(block, state);
+      if (!workList.remove(block, state)) {
+        throw new IllegalStateException();
+      }
+
+      BlockVisitor visitor = new BlockVisitor(state.dup());
+      visitor.visit(block);
+      TaintProgramState exitState = visitor.state;
+
+      callees.addAll(exitState.callees);
+
+      if (block.successors().contains(cfg.exitBlock())) {
+        if (!returnsVoid) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1066)